include/json/reader.h

```diff
@@ -337,7 +363,7 @@ class JSON_API CharReaderBuilder : public CharReader::Factory {
   CharReaderBuilder();
   ~CharReaderBuilder() override;
 
-  CharReader* newCharReader() const override;
+  CharReader::Impl* newCharReader() const override;
```

Comment:
I think I'm not understanding your suggestion about this here (https://github.com/open-source-parsers/jsoncpp/pull/1281#issuecomment-816927120)

> Close but I would not do `dynamic_cast` of `this`, but rather hold a polymorphic `_impl`. `OurCharReader` would implement a polymorphic interface and implement `getStructuredErrors()` directly.
> 
> So it would be:
> 
> ```c++
> CharReader {
> public:
>     std::vector<StructuredError> getStructuredErrors() const;
> 
>     /** apologetically virtual only for ABI compatibility. Not an extension point, so `final`. */
>     virtual bool parse(etc...) final;
> 
> private:
>     class Impl;  // Privately provides the polymorphic interface that CharReader provides today.
>     std::unique_ptr<Impl> _impl;
> };
> 
> // In cpp file to prevent inlining and control visibility.
> 
> class CharReader::Impl {
> public:
>     virtual ~Impl() = default;
>     virtual bool parse(etc...) = 0;
>     virtual std::vector<StructuredError> getStructuredErrors() const = 0;
> };
> 
> auto CharReader::getStructuredErrors() const -> std::vector<StructuredError> {
>     return _impl->getStructuredErrors();
> }
> 
> bool CharReader::parse(etc...) {
>     return _impl->parse(etc...);
> }
> ```

This particular change was part of a series of changes related to making `OurCharReader` override `CharReader::Impl`. I did this since you suggested `OurCharReader` implement `getStructuredErrors()` while `CharReader::getStructuredErrors()` calls `_impl->getStructuredErrors()`. Can you explain how you envisioned `_impl` getting the implementation of `getStructuredErrors()` in your suggestion for this?

Reply:
@BillyDonahue can you get back to me about this question so I can finish resolving your comments?

Reply:
Yes sorry for the delay.

I wasn't sure what the question meant by "override" of `CharReader::Impl`, so I was waiting until I had a good amount of time to spend on it. Okay.

I'm thinking that `newCharReader` just returns `CharReader*` as it always has.

ABI-wise, no client code depends on the internals of `CharReader`.  They only ever work with it by pointer and by calling functions on those pointers that live in the JsonCpp library.

But the `newCharReader` function can place whatever kind of `Impl` it wants into the `CharReader` it returns. This hides the details from the callers.

It would make an `Impl` subclass that has a `getStructuredErrors` appropriate for a `CharReader`, and place that into the `CharReader` to which it returns a pointer.